### DISCUSSION TITLE
RTK query feature request: Initial Query Data and Updates from Mutation Responses

### DISCUSSION THREAD

**Discussion by DouglasDev** [reduxjs/redux-toolkit#2206] [Ideas]
Hi, I've been converting a medium size app to use RTK query and I have mixed feelings about it. 90% of the time, RTK query fits my use case well and is very pleasant to work with, but the other 10% of the time, I feel like I'm trying to fit a square peg into a round hole. My main issues:

1. No easy way to set initial query data. I need to do this in the case of loading user session data from a cookie. Once I load the data from the cookie, I use a saved token to request fresh session data from the server. RTK supports "optimistic updates", which seems like would fit this use case, but it appears that I can only do an optimistic update if I already have cached data, which is impossible in this case.

2. Related to number 1,  user session data is also returned by the response to the account login and register endpoints in my api, so I need a way to add this response to the same user session state slice. It seems that a "pessimistic update" would fit this use case, but again, I cannot do the update if there is no cached data, which of course I don't have if the user isn't logged in.

Because of the above 2 issues, I needed to create another user session state slice without the use of RTK query, that is managed differently than all my other api state, entirely because the RTK query api could not accommodate my use case (which I think is fairly common). 

I later ran into another similar issue. I have a page where a user can edit some data and then save it to the server. The first time they edit it, the initial data is generated client side. Then when they make the post request to save the data, the server returns the updated data in the response. The problem is nearly identical to the one above: I need a simple way to set initial query data and then update it based on the server response.

Because of the difficulty I encountered doing this, I'm considering switching to react query which has these features built in:
https://react-query.tanstack.com/guides/initial-query-data
https://react-query.tanstack.com/guides/placeholder-query-data
https://react-query.tanstack.com/guides/updates-from-mutation-responses

Unrelated, I've gotten used to using the "state" tab in the redux dev tools, where I can see the current state at a glance. However, with the rtk query dev tools, I need to click through multiple nested objects in order to get to the `data` field for a query, meaning that I often resort to console.logging the `data` rather than using the dev tools. I realize that technically rtk query saves lots of other types of state, but most of the time, I'm only interested in the `data` that is actually being used in my app.

Lastly, there's a part of my app where I need to make multiple parallel requests to the same endpoint. I eventually figured out how to accomplish this in RTK query, but it required writing a couple of helper functions and wasn't straight forward. So perhaps this is not a missing feature per se, but either a built in helper function or an example in the docs would help a lot with this.

I hope this post doesn't come across as too critical. I appreciate the hard work of the redux developers and overall I think rtk query is an excellent library, but the two features that I mentioned would save devs a lot of frustration. Thanks!

**Comment by markerikson** [comment]
Hiya.  A couple quick thoughts:

- The "initial data" scenario _may_ be doable via the hydration support that was added in 1.7 ( https://redux-toolkit.js.org/rtk-query/usage/persistence-and-rehydration , https://redux-toolkit.js.org/rtk-query/api/createApi#extractrehydrationinfo ), although I haven't tried doing anything like that myself.  That said, just glancing at those it feels like it probably wouldn't be as simple as R-Q's option for that atm.
- I'm not sure I'm following the data flow you're describing for the user session data scenario. When you say "add this response to the same user session state slice", are you talking about an RTKQ endpoint cache entry for something like a `/getUser` endpoint?

There _is_ a draft PR for an `upsertQueryData` util method that would potentially help here, but we haven't gotten back around to looking at it: https://github.com/reduxjs/redux-toolkit/pull/2007  .  But, it _is_ something that seems like a good candidate for RTK 1.9 (per https://github.com/reduxjs/redux-toolkit/issues/2003 ).  If you get a chance, could you maybe try out the build from that PR and see if it's helpful?  Feedback on the API would be useful here.

I also haven't had time to use the RTK Query DevTools monitor much, but I'd agree that it would probably be useful to focus more on the data.  I just filed https://github.com/reduxjs/redux-devtools/issues/1126 with that suggestion.

Finally, what aspect of the "parallel requests" did you feel was confusing and not straightforward?  What did the final result look like?

**Reply by DouglasDev** [reply]
> I'm not sure I'm following the data flow you're describing for the user session data scenario. When you say "add this response to the same user session state slice", are you talking about an RTKQ endpoint cache entry for something like a /getUser endpoint?

I have a state slice/cache entry which contains user name, user id, session token, etc. This state may come from multiple places: a cookie, the response to a /register or /login POST request, or the response to a /getUser GET request. I would like to have a default state which is set from a cookie, or if there is no cookie, a different default state like `{ authenticated: false }`, and then when a request is made, update that default state with the response data. Does that make sense? Thanks for pointing out extractRehydrationInfo, however it looks like it would be quite a bit of work to get it to fit my exact use case.  I will look at that PR when I have a chance. But basically I think React Query has a good approach to solving this issue.

> Finally, what aspect of the "parallel requests" did you feel was confusing and not straightforward? What did the final result look like?

I ended up writing the follow utility function to combine the results of the queries:
```
function combineQueryResults(queries) {
  const errorQuery = queries.find((query) => query.error);
  if (errorQuery) return errorQuery;
  const loadingQuery = queries.find((query) => query.isLoading);
  if (loadingQuery) return loadingQuery;
  else return { data: queries.map((query) => query.data) };
}
```
plus this function to make the requests:
```
async queryFn(queries, _queryApi, _extraOptions, fetchWithBQ) {
  const requests = queries.map((query) =>
    fetchWithBQ(query)
  );
  const results = await Promise.all(requests);
  return combineQueryResults(results)
}
```
I guess whether this is straightforward is subjective, but it seems to me that there should be a simpler way of combining the results of multiple queries (actually looking over the code again, I'm not even sure if my implementation is entirely correct)

Anyway, thanks for your help!

**Reply by markerikson** [reply]
The phrase "state slice/cache entry" is part of what's confusing me :) Because those are two different things.

Are you actually calling `createSlice` and storing this data as its own slice, like `state.auth`?  Or is this an endpoint entry, like `endpoints: {getUser: builder.query()}`?

If it's a separate slice, then you can use `createSlice.extraReducers` to listen for endpoint requests fulfilling, as well as dispatch your own actions to update that same state in the slice.  If it's an endpoint, then yeah, that "upsert data" method is probably what you need for now.

I could at least potentially imagine that an `initialData` arg might be feasible, but I don't think anyone's specifically suggested that yet, and we haven't looked into it.

I'm still not quite following the "parallel queries" bit either.

The first intended usage would be just having multiple separate query hooks in the same component:

```js
const { data : result1 } = useFirstQuery();
const { data : result2 } = useSecondQuery();
```

Those will both immediately kick off in parallel when the component mounts.

Looking at your snippet, it seems like you're trying to do something different, but I'm not exactly sure what.  Can you show me an example of what you're trying to accomplish in the component (or wherever you want to kick off these parallel requests), as well as how you're using those util functions?

**Reply by phryneas** [reply]
As per my comment in that PR, `upsertQueryData` needs a pretty big revamp for the approach to cover all bases.

For your user auth data though: we are recommending to keep that outside of RTK Query in a normal Redux slice/httpOnly cookie or somewhere else in the first place. That would need it's own hydration with redux-persist/whichever tool would be fitting.  
If that initial data contains anything else I would suggest that you are probably overoptimizing - if that data fits into a <4kb cookie, it should not have any real impact on the application to just make that query again on application load.

And you can always just use a default initiailizer if you want initial/default data, I don't really see why that would need to be an extra feature:

```js
const { data = initialData, isLoading } = useMyQuery(arg)
```

**Reply by DouglasDev** [reply]
> The phrase "state slice/cache entry" is part of what's confusing me :) Because those are two different things.
> 
> Are you actually calling `createSlice` and storing this data as its own slice, like `state.auth`? Or is this an endpoint entry, like `endpoints: {getUser: builder.query()}`?
> 
> If it's a separate slice, then you can use `createSlice.extraReducers` to listen for endpoint requests fulfilling, as well as dispatch your own actions to update that same state in the slice. If it's an endpoint, then yeah, that "upsert data" method is probably what you need for now.

Sorry for the confusion. What I mean is that conceptually, it is like an endpoint entry, however because I need to set initial data from a cookie, I ended up using createSlice with extraReducers like you said. I feel this is not a totally satisfactory approach since now I have one bit of cached server state that works differently than all the cached server state solely because the RTK query api doesn't support initial query data.
 
> I'm still not quite following the "parallel queries" bit either.
> 
> The first intended usage would be just having multiple separate query hooks in the same component:
> 
> ```js
> const { data : result1 } = useFirstQuery();
> const { data : result2 } = useSecondQuery();
> ```
> 
> Those will both immediately kick off in parallel when the component mounts.
> 
> Looking at your snippet, it seems like you're trying to do something different, but I'm not exactly sure what. Can you show me an example of what you're trying to accomplish in the component (or wherever you want to kick off these parallel requests), as well as how you're using those util functions?

Basically there's couple of places in my app where I need to make multiple requests to an endpoint and then do something with the combined data returned from all requests. The typical way of doing this without rtk query would be something like:

```
Promise.all(requestsArray).then(responsesArray=>{
//do something with the responsesArray
}).catch(error=>{
//handle error
})
```

However queries in RTK query always return an object with `data` or `error`, so if all requests resolve, I would end up with an array of  `[{data:data1},{data:data2},{data:data3}]` when really I want something like `{data:[data1,data,2,data3]}`. Also I want to be able to throw immediately on the first error and have `isLoading` be `true` as long as there is one request that is pending, rather than have to handle the `error` and `isLoading` of each request separately. Does that all make sense?

**Reply by DouglasDev** [reply]
> For your user auth data though: we are recommending to keep that outside of RTK Query in a normal Redux slice/httpOnly cookie or somewhere else in the first place. That would need it's own hydration with redux-persist/whichever tool would be fitting. 

That is the approach I went with but for the reasons I mentioned above, I don't think it's an ideal approach.

>If that initial data contains anything else I would suggest that you are probably overoptimizing - if that data fits into a <4kb cookie, it should not have any real impact on the application to just make that query again on application load.

My issue is not related to optimization or trying to avoid making a query. It's that I need to load some initial data from a cookie before I can even make the other requests or load the app at all since the data in the cookie contains the auth token and other essential data.

> And you can always just use a default initiailizer if you want initial/default data, I don't really see why that would need to be an extra feature:
> 
> ```js
> const { data = initialData, isLoading } = useMyQuery(arg)
> ```
That is fine except RTK query doesn't allow me to perform manual cache updates without any initial cached data, which is necessary for my use case.

**Reply by phryneas** [reply]
> > For your user auth data though: we are recommending to keep that outside of RTK Query in a normal Redux slice/httpOnly cookie or somewhere else in the first place. That would need it's own hydration with redux-persist/whichever tool would be fitting.
> 
> That is the approach I went with but for the reasons I mentioned above, I don't think it's an ideal approach.

To make this clear: RTK Query keeps cache data from a request for you, nothing more. 

Something like an auth token should definitely not be "somewhere in your cache", but in a dedicated place. At no point we do or will encourage just having that value hang around in your cache, where it might get collected and thrown away, re-queried for various reasons etc. 
A cache is a cache (data that can at any point just be thrown away and rerequested) and application data is application data.

**Reply by DouglasDev** [reply]
> > > For your user auth data though: we are recommending to keep that outside of RTK Query in a normal Redux slice/httpOnly cookie or somewhere else in the first place. That would need it's own hydration with redux-persist/whichever tool would be fitting.
> > 
> > 
> > That is the approach I went with but for the reasons I mentioned above, I don't think it's an ideal approach.
> 
> To make this clear: RTK Query keeps cache data from a request for you, nothing more.
> 
> Something like an auth token should definitely not be "somewhere in your cache", but in a dedicated place. At no point we do or will encourage just having that value hang around in your cache, where it might get collected and thrown away, re-queried for various reasons etc. A cache is a cache (data that can at any point just be thrown away and rerequested) and application data is application data.

That exactly describes my use case. I just need to read the initial data from a cookie for the initial loading of my app. After that there are other endpoints that will return fresh user data (this is what is saved in the cookie) so it's totally fine if that initial data is thrown away and rerequested, and in fact that is what my app is currently doing. So I would argue that it definitely does belong in my cache and it is not client-side application data.

**Reply by phryneas** [reply]
Your authentication token is not a cache value. 
As the author of RTK Query, I can say with confidence that it was never intended to hold this kind of data and the sub-library has not been designed to hold this type of data. RTK Query is still part of RTK and a normal RTK Slice is much more suited to hold this type of data.

(Either way, since you are already using cookies and seem to have control over your server, something like an authentication token should not even be held in JavaScript in the first place, as a security best practice. It should be a `secure httpOnly` cookie, set by your server and read by your server without JavaScript ever touching it.)

**Reply by DouglasDev** [reply]
The cookie contains several other types of data in addition to the token, and they are all used for the initial loading of my app, so I think my point still stands. But I do appreciate the security tip! Didn't know about the `secure httpOnly` cookie.
